### ISSUE TITLE
[CUDA] illegal memory read on compute_alphas_and_betas_v8

### ISSUE THREAD

**Issue by kokol16** [pytorch/pytorch#189670]
### 🐛 Describe the bug

```python
import torch
print("torch:", torch.__version__)

torch.cuda.init()
torch.backends.cudnn.enabled = True
torch.backends.cudnn.deterministic = True

T, N, C = 50, 3, 15
log_probs = torch.randn(T, N, C, device='cuda', requires_grad=True).log_softmax(-1)
targets = torch.full((75,), -2**31, dtype=torch.int32)
input_lengths = torch.tensor([T, T, T], dtype=torch.int32)
target_lengths = torch.tensor([1, 1, 1], dtype=torch.int32)

loss = torch.nn.functional.ctc_loss(
    log_probs, targets, input_lengths, target_lengths,
    blank=0, reduction='sum', zero_infinity=True,
)
loss.backward()
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 4 bytes
=========     at void cudnn::compute_alphas_and_betas_v8<float>(cudnnCTCLossStruct, const T1 *, const int *, const int *, const int *, int *, int *, int *, T1 *, T1 *, T1 *, ulonglong3, ulonglong3, int)+0x4470
=========     by thread (1,0,0) in block (0,0,0)
=========     Address 0x7fa469e9e55c is out of bounds
=========     and is 8589286052 bytes before the nearest allocation at 0x7fa669e00000 of size 2097152 bytes
```

### Versions

Collecting environment information...

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @csarofeen @ptrblck @eqy @nWEIdia @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @msaroufim @tinglvv

**Comment by MDSALMANSHAMS** [comment]
`targets` here contains `-2**31`, but CTC targets must lie in `[0, num_labels)`. That negative label flows into the cuDNN alpha/beta computation (`compute_alphas_and_betas_v8`) and indexes `log_probs` out of bounds. Validating that all `targets` are non-negative and `< num_classes` before the cuDNN CTC call would catch this cleanly. Would a PR adding that input check be welcome, and could this be marked `actionable`?

**Comment by ngimel** [comment]
Make sure your PR doesn't introduce host-to-device sync and does device-side assert instead

**Comment by ngimel** [comment]
@eqy looks like a bug in cudnn, cudnn itself should be doing device-side assert here

**Comment by MDSALMANSHAMS** [comment]
@ngimel @eqy — following up on the "cudnn should be doing the device-side assert" point, since it decides the scope here. I agree the OOB read happens inside cuDNN's alpha/beta kernel and we can't put an assert in there. But in the path this repro takes, the offending value is still on the **host** at the moment we hand it over, so a check costs no sync:

In `_cudnn_ctc_loss_tensor` (`aten/src/ATen/native/cudnn/LossCTC.cpp`), `targets_t` is copied to the device by us:

```cpp
  Tensor targets_t_ = targets_t;
  ...
  if (targets_t.device().type() == at::kCPU) {
    targets_t_ = targets_t.to(Device(at::kCUDA));   // L266 on main
  }
```

Validating `0 <= t < log_probs.size(2)` *before* that copy is a plain host-side read of a CPU tensor — no host-to-device sync, so it doesn't run into the rule you flagged. The legacy `_cudnn_ctc_loss` entry point is even simpler: it requires CPU targets outright (`checkBackend(c, {*targets}, Backend::CPU)`), so the same check is free there too.

That leaves only the case where the caller passes targets already on the GPU, which would genuinely need either a device-side assert in a small pre-pass or to be left to cuDNN.

So the question before I put up a fix: **is a host-side validation on the CPU-targets paths (both entry points) the scope you want, leaving GPU-resident targets to cuDNN — or would you rather this be handled entirely on the cuDNN side and nothing land here?** Happy either way, I just don't want to send a PR against the direction you two settle on.

Separately, while reading that function I think there is an unrelated one-line bug just below the snippet above:

```cpp
  if (input_lengths.device().type() == at::kCPU) {
    input_lengths_ = input_lengths.to(Device(at::kCUDA));
  }
  if (input_lengths.device().type() == at::kCPU) {     // <- should be target_lengths
    target_lengths_ = target_lengths.to(Device(at::kCUDA));
  }
```

The second condition tests `input_lengths` but assigns `target_lengths_`. If a caller passes `input_lengths` on CUDA and `target_lengths` on CPU, `target_lengths_` is never moved, and `target_lengths_.data_ptr<int>()` is then passed straight to `cudnnCTCLoss_v8` as a device pointer. Happy to fold the one-character fix into whatever lands here, or split it out if you prefer.

**Comment by eqy** [comment]
cuDNN will be adding the fix

**Comment by MDSALMANSHAMS** [comment]
Thanks @eqy — that settles the targets check, I'll leave it to cuDNN.

One thing from the tail of my last comment that's independent of the cuDNN fix, in case it got lost: `aten/src/ATen/native/cudnn/LossCTC.cpp` L268-273 has a copy-paste slip — the second `if` tests `input_lengths` but assigns `target_lengths_`:

```cpp
  if (input_lengths.device().type() == at::kCPU) {
    input_lengths_ = input_lengths.to(Device(at::kCUDA));
  }
  if (input_lengths.device().type() == at::kCPU) {   // <- should be target_lengths
    target_lengths_ = target_lengths.to(Device(at::kCUDA));
  }
```

So with `input_lengths` on CUDA and `target_lengths` on CPU, `target_lengths_` never gets moved and its pointer goes to `cudnnCTCLoss_v8` (L338) as device memory.

Happy to put up the one-word fix if you want it.

**Comment by malfet** [comment]
Duplicate of #195547. This is one instance of a class of missing shape/size checks that lead to out-of-bounds tensor access — here surfaced by compute-sanitizer on a CUDA kernel rather than by ASAN on CPU, but the same missing bounds contract. The class is now tracked under the umbrella issue #195547, which also explains why per-operator TORCH_CHECK PRs are not the fix being sought. Please follow #195547.
